File: frontend/src/lib/supabase.ts
/**
 * Supabase Client Configuration
 * 
 * Frontend Supabase client initialization and OAuth provider management
 * for the Duolingo clone authentication system.
 */

import { createClient, SupabaseClient, AuthSession, User } from '@supabase/supabase-js'

// Environment variables for Supabase configuration
const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL!
const supabaseAnonKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!

// Create Supabase client with enhanced configuration
export const supabase: SupabaseClient = createClient(supabaseUrl, supabaseAnonKey, {
  auth: {
    autoRefreshToken: true,
    persistSession: true,
    detectSessionInUrl: true,
    flowType: 'pkce' // Use PKCE for enhanced security
  },
  realtime: {
    params: {
      eventsPerSecond: 10
    }
  }
})

// Types for OAuth providers
export type OAuthProvider = 'google' | 'apple' | 'facebook' | 'github' | 'twitter'

export interface OAuthOptions {
  provider: OAuthProvider
  redirectTo?: string
  scopes?: string
  queryParams?: Record<string, string>
}

export interface AuthState {
  session: AuthSession | null
  user: User | null
  loading: boolean
  error: string | null
}

export interface SignUpData {
  email: string
  password: string
  firstName?: string
  lastName?: string
  metadata?: Record<string, unknown>
}

export interface SignInData {
  email: string
  password: string
  rememberMe?: boolean
}

export interface PasswordResetData {
  email: string
  redirectTo?: string
}

/**
 * Authentication utilities class for frontend Supabase operations
 */
export class SupabaseAuth {
  private client: SupabaseClient

  constructor(client: SupabaseClient) {
    this.client = client
  }

  /**
   * Sign up with email and password
   */
  async signUp(data: SignUpData) {
    try {
      const { user, session, error } = await this.client.auth.signUp({
        email: data.email,
        password: data.password,
        options: {
          data: {
            first_name: data.firstName,
            last_name: data.lastName,
            ...data.metadata
          }
        }
      })

      if (error) throw error

      return { user, session, error: null }
    } catch (error) {
      console.error('Sign up error:', error)
      return { 
        user: null, 
        session: null, 
        error: error instanceof Error ? error.message : 'Sign up failed' 
      }
    }
  }

  /**
   * Sign in with email and password
   */
  async signIn(data: SignInData) {
    try {
      const { user, session, error } = await this.client.auth.signInWithPassword({
        email: data.email,
        password: data.password
      })

      if (error) throw error

      // Handle "Remember Me" functionality by extending session duration
      if (data.rememberMe && session) {
        // Note: Session duration is handled by Supabase configuration
        // This could be enhanced with custom session management if needed
      }

      return { user, session, error: null }
    } catch (error) {
      console.error('Sign in error:', error)
      return { 
        user: null, 
        session: null, 
        error: error instanceof Error ? error.message : 'Sign in failed' 
      }
    }
  }

  /**
   * Sign in with OAuth provider
   */
  async signInWithOAuth(options: OAuthOptions) {
    try {
      const { data, error } = await this.client.auth.signInWithOAuth({
        provider: options.provider,
        options: {
          redirectTo: options.redirectTo || `${window.location.origin}/auth/callback`,
          scopes: options.scopes || this.getDefaultScopes(options.provider),
          queryParams: options.queryParams
        }
      })

      if (error) throw error

      return { data, error: null }
    } catch (error) {
      console.error('OAuth sign in error:', error)
      return { 
        data: null, 
        error: error instanceof Error ? error.message : 'OAuth sign in failed' 
      }
    }
  }

  /**
   * Sign out current user
   */
  async signOut() {
    try {
      const { error } = await this.client.auth.signOut()
      if (error) throw error
      return { error: null }
    } catch (error) {
      console.error('Sign out error:', error)
      return { error: error instanceof Error ? error.message : 'Sign out failed' }
    }
  }

  /**
   * Reset password
   */
  async resetPassword(data: PasswordResetData) {
    try {
      const { error } = await this.client.auth.resetPasswordForEmail(
        data.email,
        {
          redirectTo: data.redirectTo || `${window.location.origin}/auth/reset-password`
        }
      )

      if (error) throw error
      return { error: null }
    } catch (error) {
      console.error('Password reset error:', error)
      return { error: error instanceof Error ? error.message : 'Password reset failed' }
    }
  }

  /**
   * Update password
   */
  async updatePassword(newPassword: string) {
    try {
      const { error } = await this.client.auth.updateUser({
        password: newPassword
      })

      if (error) throw error
      return { error: null }
    } catch (error) {
      console.error('Password update error:', error)
      return { error: error instanceof Error ? error.message : 'Password update failed' }
    }
  }

  /**
   * Get current session
   */
  async getSession() {
    try {
      const { data: { session }, error } = await this.client.auth.getSession()
      if (error) throw error
      return { session, error: null }
    } catch (error) {
      console.error('Get session error:', error)
      return { session: null, error: error instanceof Error ? error.message : 'Failed to get session' }
    }
  }

  /**
   * Get current user
   */
  async getUser() {
    try {
      const { data: { user }, error } = await this.client.auth.getUser()
      if (error) throw error
      return { user, error: null }
    } catch (error) {
      console.error('Get user error:', error)
      return { user: null, error: error instanceof Error ? error.message : 'Failed to get user' }
    }
  }

  /**
   * Refresh session
   */
  async refreshSession() {
    try {
      const { data, error } = await this.client.auth.refreshSession()
      if (error) throw error
      return { session: data.session, user: data.user, error: null }
    } catch (error) {
      console.error('Refresh session error:', error)
      return { 
        session: null, 
        user: null, 
        error: error instanceof Error ? error.message : 'Failed to refresh session' 
      }
    }
  }

  /**
   * Listen to auth state changes
   */
  onAuthStateChange(callback: (event: string, session: AuthSession | null) => void) {
    return this.client.auth.onAuthStateChange(callback)
  }

  /**
   * Get default scopes for OAuth providers
   */
  private getDefaultScopes(provider: OAuthProvider): string {
    const scopesMap: Record<OAuthProvider, string> = {
      google: 'openid email profile',
      facebook: 'email public_profile',
      apple: 'email name',
      github: 'user:email',
      twitter: 'users.read tweet.read'
    }

    return scopesMap[provider] || 'openid email profile'
  }
}

// Export configured auth instance
export const auth = new SupabaseAuth(supabase)

/**
 * Helper function to check if user is authenticated
 */
export const isAuthenticated = async (): Promise<boolean> => {
  const { session } = await auth.getSession()
  return !!session
}

/**
 * Helper function to get user profile from Supabase
 */
export const getUserProfile = async (userId: string) => {
  try {
    const { data, error } = await supabase
      .from('profiles')
      .select('*')
      .eq('user_id', userId)
      .single()

    if (error) throw error
    return { profile: data, error: null }
  } catch (error) {
    console.error('Get user profile error:', error)
    return { 
      profile: null, 
      error: error instanceof Error ? error.message : 'Failed to get user profile' 
    }
  }
}

/**
 * Helper function to update user profile in Supabase
 */
export const updateUserProfile = async (userId: string, updates: Record<string, unknown>) => {
  try {
    const { data, error } = await supabase
      .from('profiles')
      .update(updates)
      .eq('user_id', userId)
      .select()
      .single()

    if (error) throw error
    return { profile: data, error: null }
  } catch (error) {
    console.error('Update user profile error:', error)
    return { 
      profile: null, 
      error: error instanceof Error ? error.message : 'Failed to update user profile' 
    }
  }
}

/**
 * Helper function to create user profile in Supabase
 */
export const createUserProfile = async (userId: string, profileData: Record<string, unknown>) => {
  try {
    const { data, error } = await supabase
      .from('profiles')
      .insert({
        user_id: userId,
        ...profileData
      })
      .select()
      .single()

    if (error) throw error
    return { profile: data, error: null }
  } catch (error) {
    console.error('Create user profile error:', error)
    return { 
      profile: null, 
      error: error instanceof Error ? error.message : 'Failed to create user profile' 
    }
  }
}

// Environment configuration validation
if (!supabaseUrl || !supabaseAnonKey) {
  console.error('Missing Supabase environment variables. Please check your .env.local file.')
}

File: backend/app/core/supabase.py
"""
Supabase Client Configuration

Handles Supabase client initialization and OAuth provider management
for the Duolingo clone authentication system.
"""

import logging
from typing import Optional, Dict, Any
from supabase import create_client, Client
from supabase.client import ClientOptions
from gotrue.errors import AuthError
from app.core.config import get_settings

logger = logging.getLogger(__name__)


class SupabaseClient:
    """
    Supabase client wrapper for authentication and user management.
    
    Provides a singleton pattern for Supabase client access and includes
    helper methods for OAuth provider management and user operations.
    """
    
    _instance: Optional['SupabaseClient'] = None
    _client: Optional[Client] = None
    
    def __new__(cls) -> 'SupabaseClient':
        """Ensure singleton pattern for Supabase client."""
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance
    
    def __init__(self):
        """Initialize Supabase client if not already initialized."""
        if self._client is None:
            self._initialize_client()
    
    def _initialize_client(self) -> None:
        """Initialize the Supabase client with configuration."""
        settings = get_settings()
        
        if not settings.has_supabase_config:
            logger.warning("Supabase configuration incomplete. Client will not be initialized.")
            return
        
        try:
            # Configure client options for enhanced security
            options = ClientOptions(
                auto_refresh_token=True,
                persist_session=True,
                detect_session_in_url=False,  # We handle OAuth callbacks manually
                flow_type="pkce"  # Use PKCE for enhanced security
            )
            
            self._client = create_client(
                supabase_url=settings.supabase_url,
                supabase_key=settings.supabase_anon_key,
                options=options
            )
            
            logger.info("Supabase client initialized successfully")
            
        except Exception as e:
            logger.error(f"Failed to initialize Supabase client: {str(e)}")
            raise
    
    @property
    def client(self) -> Optional[Client]:
        """Get the Supabase client instance."""
        return self._client
    
    @property
    def auth(self):
        """Get the Supabase auth client."""
        if not self._client:
            raise RuntimeError("Supabase client not initialized")
        return self._client.auth
    
    @property
    def db(self):
        """Get the Supabase database client."""
        if not self._client:
            raise RuntimeError("Supabase client not initialized")
        return self._client.table
    
    def is_configured(self) -> bool:
        """Check if Supabase client is properly configured."""
        return self._client is not None
    
    async def sign_in_with_oauth(self, provider: str, redirect_url: Optional[str] = None) -> Dict[str, Any]:
        """
        Initiate OAuth sign-in flow with specified provider.
        
        Args:
            provider: OAuth provider name (google, apple, facebook, tiktok)
            redirect_url: Custom redirect URL (defaults to configured OAuth callback)
            
        Returns:
            Dictionary containing OAuth URL and session information
            
        Raises:
            AuthError: If OAuth initiation fails
            RuntimeError: If client not configured
        """
        if not self._client:
            raise RuntimeError("Supabase client not initialized")
        
        settings = get_settings()
        callback_url = redirect_url or settings.oauth_callback_url
        
        try:
            response = await self.auth.sign_in_with_oauth({
                "provider": provider,
                "options": {
                    "redirect_to": callback_url,
                    "scopes": self._get_provider_scopes(provider)
                }
            })
            
            logger.info(f"OAuth sign-in initiated for provider: {provider}")
            return response
            
        except AuthError as e:
            logger.error(f"OAuth sign-in failed for provider {provider}: {str(e)}")
            raise
    
    async def handle_oauth_callback(self, code: str, state: str) -> Dict[str, Any]:
        """
        Handle OAuth callback and exchange code for session.
        
        Args:
            code: Authorization code from OAuth provider
            state: State parameter for CSRF protection
            
        Returns:
            Dictionary containing user session and profile information
            
        Raises:
            AuthError: If callback handling fails
        """
        if not self._client:
            raise RuntimeError("Supabase client not initialized")
        
        try:
            response = await self.auth.exchange_code_for_session({
                "auth_code": code,
                "code_verifier": state  # PKCE code verifier
            })
            
            logger.info("OAuth callback handled successfully")
            return response
            
        except AuthError as e:
            logger.error(f"OAuth callback handling failed: {str(e)}")
            raise
    
    async def sign_out(self) -> None:
        """Sign out current user and invalidate session."""
        if not self._client:
            raise RuntimeError("Supabase client not initialized")
        
        try:
            await self.auth.sign_out()
            logger.info("User signed out successfully")
            
        except AuthError as e:
            logger.error(f"Sign out failed: {str(e)}")
            raise
    
    async def get_user(self) -> Optional[Dict[str, Any]]:
        """
        Get current authenticated user information.
        
        Returns:
            User dictionary if authenticated, None otherwise
        """
        if not self._client:
            return None
        
        try:
            user = await self.auth.get_user()
            return user.user if user else None
            
        except AuthError:
            return None
    
    async def refresh_session(self) -> Optional[Dict[str, Any]]:
        """
        Refresh current user session.
        
        Returns:
            Refreshed session information or None if refresh fails
        """
        if not self._client:
            return None
        
        try:
            response = await self.auth.refresh_session()
            return response
            
        except AuthError as e:
            logger.error(f"Session refresh failed: {str(e)}")
            return None
    
    def _get_provider_scopes(self, provider: str) -> str:
        """
        Get appropriate scopes for OAuth provider.
        
        Args:
            provider: OAuth provider name
            
        Returns:
            Space-separated string of scopes
        """
        scopes_map = {
            "google": "openid email profile",
            "facebook": "email public_profile",
            "apple": "email name",
            "tiktok": "user.info.basic user.info.profile"
        }
        
        return scopes_map.get(provider, "openid email profile")
    
    async def create_user_profile(self, user_id: str, profile_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Create user profile in application database.
        
        Args:
            user_id: Supabase user ID
            profile_data: User profile information
            
        Returns:
            Created profile data
        """
        if not self._client:
            raise RuntimeError("Supabase client not initialized")
        
        try:
            response = await self.db("profiles").insert({
                "user_id": user_id,
                **profile_data
            }).execute()
            
            logger.info(f"User profile created for user: {user_id}")
            return response.data[0] if response.data else {}
            
        except Exception as e:
            logger.error(f"Failed to create user profile: {str(e)}")
            raise
    
    async def get_user_profile(self, user_id: str) -> Optional[Dict[str, Any]]:
        """
        Get user profile from application database.
        
        Args:
            user_id: Supabase user ID
            
        Returns:
            User profile data or None if not found
        """
        if not self._client:
            return None
        
        try:
            response = await self.db("profiles").select("*").eq("user_id", user_id).execute()
            return response.data[0] if response.data else None
            
        except Exception as e:
            logger.error(f"Failed to get user profile: {str(e)}")
            return None


# Global Supabase client instance
_supabase_client = None


def get_supabase_client() -> SupabaseClient:
    """
    Get global Supabase client instance.
    
    Returns:
        Singleton SupabaseClient instance
    """
    global _supabase_client
    if _supabase_client is None:
        _supabase_client = SupabaseClient()
    return _supabase_client


def init_supabase() -> None:
    """
    Initialize Supabase client on application startup.
    
    This function should be called during FastAPI application startup
    to ensure Supabase client is properly configured.
    """
    client = get_supabase_client()
    if client.is_configured():
        logger.info("Supabase client configuration verified")
    else:
        logger.warning("Supabase client configuration incomplete")